[Flex] Try to use cached logical height for flex item's hypothetical cross size.

sammygill opened the pull request:
#### 8e50dd5daa89220c06096dc5961f86e225c165f7
<pre>
[Flex] Try to use cached logical height for flex item&apos;s hypothetical cross size.
<a href="https://rdar.apple.com/132034177">rdar://132034177</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=276796">https://bugs.webkit.org/show_bug.cgi?id=276796</a>

Reviewed by Alan Baradlay.

Currently, in RenderFlexibleBox::layoutAndPlaceFlexItems, we currently
call setNeedsLayout on a flex item if it has a relative logical height
(e.g. % min-height). In some cases, this can result in us doing some extra
work on subsequent layouts when we may not need to. Consider the
following piece of content:

&lt;div style=&quot;display: flex;&quot;&gt;
  &lt;some-potentially-expensive-flex-item style=&quot;height: 100%;&quot;&gt;&lt;/some-potentially-expensive-flex-item&gt;
  &lt;div&gt;foobar&lt;/div&gt;
&lt;/div&gt;

In this example, when we run flex layout, we do the following for each of
the flex items:

1.) Compute the main axis size
2.) Compute the hypothetical cross size
3.) Perform stretch alignment

For this type of content, we need to call layout() on each of the flex
items during both 2 and 3, and we should also note that calling layout on
&lt;some-potentially-expensive-flex-item&gt; may be &quot;expensive,&quot; for a variety
of reasons. On a subsequent layout, for example, if we change the content
inside the &lt;div&gt;, we end up calling setNeedsLayout() on the first flex
item due to its % logical height and perform layout on it in order to
compute its hypothetical cross axis size again. During this process, we
unfortunately end up losing the item&apos;s stretched size and so must
stretch it again, which requires calling layout() on it one final time.

This patch attempts to mitigate this less than ideal behavior by
attempting to reuse a previously stored hypothetical cross axis size on
a subsequent layout. Rather than invalidating the flex item and
attempting to recompute it due to the % logical height, we can use the
previous value because the flexbox itself has an indefinite logical
height. This causes the flex items&apos; % logical height to behave as auto,
and the resulting size will be based just off of its content.

RenderFlexibleBox already seems to cache the logical heights for its
flex items within a map called m_intrinsicContentLogicalHeights. This
cache is populated when a renderer goes through layout as one of its
first steps inside of updateLogicalHeight(). Inside of
RenderFlexibleBox::layoutAndPlaceItems: we will try not to invalidate the
flex item if the following are true for flex items with a relative
logical height:

1.) The main axis direction is the same as the flex item&apos;s inline direction
2.) The flexbox has an indefinite logical height
3.) We have a previously cached value for the cross size

Note that if the flex item&apos;s content may have been mutated, we will still
call layout on it due to the mutation, but during layout, we will end up
updating the cached value.

When the flexbox code needs the cross size of the flex item, for example,
to determine the size of the line, it will use
crossAxisIntrinsicExtentForFlexItem, which checks to see if it has a
cached cross size to use.

In order to use this cached size correctly we also needed to add some
extra invalidation when certain types of mutations occur. In this patch
we started removing percent logical height items from the cache in two
scenarios:

1.) When the computed logical height of the flexbox changes
2.) When a flexbox goes from being stretched to no longer stretching in
cases where itself participates in a flex or grid formatting context.

In order to accomplish 2 we allow RenderFlexibleBox to keep track of a
new bit that represents this type of damage. This bit will be set by
callers, in this case RenderFlexibleBox and RenderGrid, to inform the
flexbox that it is no longer being stretched. During flex layout it will
check to see if this bit is set and prune from the cache if it is.

On my M4 Max the included performance test went from ~590 runs/s to ~87K
runs/s.

I also included a new testcase which attempts to exercise this logic to
make sure we are computing the correct sizes and invalidating the cache
properly. This testcase includes a variety of subtests in which the
content is mutated in different ways.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/percent-height-flex-items-cross-sizes-with-mutations-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/percent-height-flex-items-cross-sizes-with-mutations.html: Added.
* PerformanceTests/Layout/flexbox-percent-height-cross-size-caching.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::layoutBlock):
(WebCore::RenderFlexibleBox::removePercentLogicalHeightFlexItemsFromIntrinsicLogicalHeightCache):
(WebCore::RenderFlexibleBox::layoutAndPlaceFlexItems):
(WebCore::RenderFlexibleBox::clearNeedsLayout):
(WebCore::RenderFlexibleBox::setNoLongerStretching):
(WebCore::updateFlexItemDirtyBitsBeforeLayout): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::styleDidChange):

Canonical link: <a href="https://commits.webkit.org/292314@main">https://commits.webkit.org/292314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5ae2d2e503e53fddf181ac811757b03b3809aa2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95296 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100341 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72668 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29941 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98299 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11409 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86019 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53063 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3759 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45136 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81230 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102378 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22344 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16304 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81666 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22592 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81066 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20370 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25637 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3036 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15597 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22314 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27450 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21973 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25447 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23712 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1291,6 +1282,20 @@ LayoutUnit RenderFlexibleBox::computeFlexBaseSizeForFlexItem(RenderBox& flexItem
+        if (flexItem.hasRelativeLogicalHeight()) {
+            auto computedLogicalHeight = flexItem.computeLogicalHeightWithoutLayout();
+            if (computedLogicalHeight != flexItem.logicalHeight())
+                flexItem.setNeedsLayout(MarkOnlyThis);
```
I think this needs some comment (and the commit message should also be more specific). I am not sure how much we can trust in a precent (or calc) height value resolve right after calling  resetLogicalHeightBeforeLayoutIfNeeded(); on the containing block. Could you explain this a bit, please?

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2271,7 +2260,22 @@ void RenderFlexibleBox::layoutAndPlaceFlexItems(LayoutUnit& crossAxisOffset, Fle
+            // percentage logical height will behave as auto. If that is the case we may
+            // have a previously cached logical height that will end up being used
+            // through crossAxisIntrinsicExtentForFlexItem.
+            if (mainAxisIsFlexItemInlineAxis(flexItem))
+                return hasDefiniteLogicalHeight() || !m_intrinsicContentLogicalHeights.contains(flexItem);
```
Do you mind adding a ref test to cover this?

sammygill committed:
[Flex] Try to use cached logical height for flex item's hypothetical cross size.
rdar://132034177
https://bugs.webkit.org/show_bug.cgi?id=276796

Reviewed by Alan Baradlay.

Currently, in RenderFlexibleBox::layoutAndPlaceFlexItems, we currently
call setNeedsLayout on a flex item if it has a relative logical height
(e.g. % min-height). In some cases, this can result in us doing some extra
work on subsequent layouts when we may not need to. Consider the
following piece of content:

<div style="display: flex;">
  <some-potentially-expensive-flex-item style="height: 100%;"></some-potentially-expensive-flex-item>
  <div>foobar</div>
</div>

In this example, when we run flex layout, we do the following for each of
the flex items:

1.) Compute the main axis size
2.) Compute the hypothetical cross size
3.) Perform stretch alignment

For this type of content, we need to call layout() on each of the flex
items during both 2 and 3, and we should also note that calling layout on
<some-potentially-expensive-flex-item> may be "expensive," for a variety
of reasons. On a subsequent layout, for example, if we change the content
inside the <div>, we end up calling setNeedsLayout() on the first flex
item due to its % logical height and perform layout on it in order to
compute its hypothetical cross axis size again. During this process, we
unfortunately end up losing the item's stretched size and so must
stretch it again, which requires calling layout() on it one final time.

This patch attempts to mitigate this less than ideal behavior by
attempting to reuse a previously stored hypothetical cross axis size on
a subsequent layout. Rather than invalidating the flex item and
attempting to recompute it due to the % logical height, we can use the
previous value because the flexbox itself has an indefinite logical
height. This causes the flex items' % logical height to behave as auto,
and the resulting size will be based just off of its content.

RenderFlexibleBox already seems to cache the logical heights for its
flex items within a map called m_intrinsicContentLogicalHeights. This
cache is populated when a renderer goes through layout as one of its
first steps inside of updateLogicalHeight(). Inside of
RenderFlexibleBox::layoutAndPlaceItems: we will try not to invalidate the
flex item if the following are true for flex items with a relative
logical height:

1.) The main axis direction is the same as the flex item's inline direction
2.) The flexbox has an indefinite logical height
3.) We have a previously cached value for the cross size

Note that if the flex item's content may have been mutated, we will still
call layout on it due to the mutation, but during layout, we will end up
updating the cached value.

When the flexbox code needs the cross size of the flex item, for example,
to determine the size of the line, it will use
crossAxisIntrinsicExtentForFlexItem, which checks to see if it has a
cached cross size to use.

In order to use this cached size correctly we also needed to add some
extra invalidation when certain types of mutations occur. In this patch
we started removing percent logical height items from the cache in two
scenarios:

1.) When the computed logical height of the flexbox changes
2.) When a flexbox goes from being stretched to no longer stretching in
cases where itself participates in a flex or grid formatting context.

In order to accomplish 2 we allow RenderFlexibleBox to keep track of a
new bit that represents this type of damage. This bit will be set by
callers, in this case RenderFlexibleBox and RenderGrid, to inform the
flexbox that it is no longer being stretched. During flex layout it will
check to see if this bit is set and prune from the cache if it is.

On my M4 Max the included performance test went from ~590 runs/s to ~87K
runs/s.

I also included a new testcase which attempts to exercise this logic to
make sure we are computing the correct sizes and invalidating the cache
properly. This testcase includes a variety of subtests in which the
content is mutated in different ways.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/percent-height-flex-items-cross-sizes-with-mutations-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/percent-height-flex-items-cross-sizes-with-mutations.html: Added.
* PerformanceTests/Layout/flexbox-percent-height-cross-size-caching.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::layoutBlock):
(WebCore::RenderFlexibleBox::removePercentLogicalHeightFlexItemsFromIntrinsicLogicalHeightCache):
(WebCore::RenderFlexibleBox::layoutAndPlaceFlexItems):
(WebCore::RenderFlexibleBox::clearNeedsLayout):
(WebCore::RenderFlexibleBox::setNoLongerStretching):
(WebCore::updateFlexItemDirtyBitsBeforeLayout): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::styleDidChange):

Canonical link: https://commits.webkit.org/292314@main
